core/src/family.cc

```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
 
 template <typename T>
 std::vector<MetricFamily> Family<T>::Collect() {
-  std::lock_guard<std::mutex> lock{mutex_};
   auto family = MetricFamily{};
-  family.name = name_;
-  family.help = help_;
-  family.type = T::metric_type;
-  for (const auto& m : metrics_) {
-    family.metric.push_back(std::move(CollectMetric(m.first, m.second.get())));
+  std::vector<std::shared_ptr<T>> to_be_removed;
+  {
+    std::lock_guard<std::mutex> lock{mutex_};
+    family.name = name_;
+    family.help = help_;
+    family.type = T::metric_type;
+    for (const auto& m : metrics_) {
+      if (!m.second->Expired()) {
+        family.metric.push_back(std::move(CollectMetric(m.first, m.second)));
+      } else if (retention_behavior_ == RetentionBehavior::Remove) {
+        to_be_removed.push_back(m.second);
+      }
+    }
+  } // end mutex
+  for (auto metric: to_be_removed) {
+    Remove(metric);
+  }
+  if (family.metric.empty()) {
+    return {};
+  } else {
+    return {family};
+  }
+}
+
+template <typename T>
+bool Family<T>::UpdateRetentionTime(const double& retention_time, const std::string& re_name, const std::map<std::string, std::string>& re_labels, const bool& bump, const bool& debug) {
+  /* Setup */
+  bool modified(false);
+  const boost::regex name_expr(re_name);
+  if (debug) {  
+    printf("UpdateRetentionTime: %s \n", re_name.c_str());
+  }
+  /* Check if the name is a match */
+  if (boost::regex_match(name_, name_expr)) {
+    for (auto metric: metrics_) {
+      bool matched(true);
+      auto hash = labels_reverse_lookup_.at(metric.second);
+      for (const auto& re_label: re_labels) {
+        /* Setup */
+        bool found(false);
+        const boost::regex key_expr(re_label.first);
+        const boost::regex value_expr(re_label.second);        
+        /* Check constant labels */
+        for (const auto& label: constant_labels_) {
+          if (boost::regex_match(label.first, key_expr) && boost::regex_match(label.second, value_expr)) {
+            found = true;
+            break;
+          }
+        }
+        /* Check metric labels */
+        for (const auto& label: labels_.at(hash)) {
+          if (boost::regex_match(label.first, key_expr) && boost::regex_match(label.second, value_expr)) {
+            found = true;
+            break;
+          }
+        }
+        /* Update matched status (matched = true if all labels up to this point are a match) */
+        if (found) {
+          matched = true;
+        } else {
+          matched = false;
+          break;
+        }
```

Comment:
```suggestion
        matched = found;
        if (!matched) break;
```

Reply:
There is sort of a bug in this logic. It should only update matched if not found.